2.0/m_joinpartspam.cpp

```diff
@@ -0,0 +1,243 @@
+	JoinPartSpam(Module* Creator) : ModeHandler(Creator, "joinpartspam", 'V', PARAM_SETONLY, MODETYPE_CHANNEL)
+		, ext("joinpartspam", Creator)
+	{
+	}
+
+	ModeAction OnModeChange(User* source, User* dest, Channel* chan, std::string& parameter, bool adding)
+	{
+		if (adding)
+		{
+			std::string::size_type colon = parameter.find(':');
+			if (colon == std::string::npos || parameter.find('-') != std::string::npos)
+			{
+				source->WriteNumeric(608, "%s %s :Invalid joinpartspam parameter", source->nick.c_str(), chan->name.c_str());
+				return MODEACTION_DENY;
+			}
+
+			unsigned int ncycles = ConvToInt(parameter.substr(0, colon));
+			unsigned int nsecs = ConvToInt(parameter.substr(colon+1));
+
+			if (ncycles < 2 || nsecs < 1)
+			{
+				source->WriteNumeric(608, "%s %s :Invalid joinpartspam parameter", source->nick.c_str(), chan->name.c_str());
+				return MODEACTION_DENY;
+			}
+
+			joinpartspamsettings* jpss = ext.get(chan);
+			if (jpss && ncycles == jpss->cycles && nsecs == jpss->secs)
+				return MODEACTION_DENY;
+
+			ext.set(chan, new joinpartspamsettings(ncycles, nsecs));
+			parameter = ConvToStr(ncycles) + ":" + ConvToStr(nsecs);
+			chan->SetModeParam(GetModeChar(), parameter);
+			return MODEACTION_ALLOW;
+		}
+		else
+		{
+			if (!chan->IsModeSet(GetModeChar()))
+				return MODEACTION_DENY;
+
+			ext.unset(chan);
+			chan->SetModeParam(GetModeChar(), "");
+			return MODEACTION_ALLOW;
+		}
+	}
+};
+
+class ModuleJoinPartSpam : public Module
+{
+	JoinPartSpam jps;
+
+ public:
+	ModuleJoinPartSpam()
+		: jps(this)
+	{
+	}
+
+	void init()
+	{
+		ServerInstance->Modules->AddService(jps);
+		ServerInstance->Modules->AddService(jps.ext);
+		Implementation eventlist[] = { I_OnUserPreJoin, I_OnUserJoin };
+		ServerInstance->Modules->Attach(eventlist, this, sizeof(eventlist)/sizeof(Implementation));
+	}
+
+	void Prioritize()
+	{
+		/* Let bans, etc. stop the join first */
+		ServerInstance->Modules->SetPriority(this, I_OnUserPreJoin, PRIORITY_LAST);
+		ServerInstance->Modules->SetPriority(this, I_OnUserJoin, PRIORITY_LAST);
+	}
+
+	/* Stop the join and clear the user's counter if they've hit the limit */
+	ModResult OnUserPreJoin(User* user, Channel* chan, const char* cname, std::string& privs, const std::string& keygiven)
+	{
+		if (!chan || !chan->IsModeSet(jps.GetModeChar()))
+			return MOD_RES_PASSTHRU;
+		if (IS_OPER(user))
+			return MOD_RES_PASSTHRU;
+
+		joinpartspamsettings* jpss = jps.ext.get(chan);
+		if (jpss)
+		{
+			const std::string& mask(user->MakeHost());
+			if (jpss->zapme(mask))
+			{
+				std::vector<std::string> parameters;
+				parameters.push_back(chan->name);
+				parameters.push_back("+b");
+				parameters.push_back(user->MakeWildHost());
+				ServerInstance->SendGlobalMode(parameters, ServerInstance->FakeClient);
+
+				user->WriteNumeric(474, "%s %s :Channel JoinPart Spam triggered (limit is %u cycles in %u secs)", user->nick.c_str(), chan->name.c_str(), jpss->cycles, jpss->secs);
```

Comment:
Can you reword this to say "join/part spam" to look more natutal instead of "JoinPart Spam"?